modules/chain-indexer/src/main/scala/org/ergoplatform/uexplorer/indexer/ChainIndexer.scala

```diff
@@ -1,99 +1,68 @@
-            Future(killSwitch.shutdown()).map(_ => Done)
-          }
-
-          val initializationF =
-            for {
-              blockHttpClient <- BlockHttpClient.withNodePoolBackend
-              pluginManager   <- PluginManager.initialize
-              graphBackendOpt <- Future.fromTry(GraphBackend(conf.graphBackendType))
-              storage         <- Future.fromTry(MvStorage.withDefaultDir(conf.mvStore.cacheSize))
-              backend         <- Future.fromTry(Backend(conf.backendType))
-              storageService = StorageService(storage, conf.mvStore)
-              blockReader    = new BlockReader(blockHttpClient)
-              blockWriter    = new BlockWriter(storage, storageService, conf.mvStore, backend, graphBackendOpt)
-              chainIndexer   = new StreamExecutor(true, blockHttpClient, blockReader, blockWriter, storage)
-              mempoolSyncer  = new MempoolSyncer(blockHttpClient)
-              initializer    = new Initializer(storage, backend, graphBackendOpt)
-              scheduler      = new Scheduler(pluginManager, chainIndexer, mempoolSyncer, initializer)
-              done <- scheduler.validateAndSchedule(0.seconds, conf.mvStore.maxIdleCompactTime + 5.seconds)
-            } yield done
-
-          initializationF.andThen {
-            case Failure(ex) =>
-              val sw = new StringWriter()
-              ex.printStackTrace(new PrintWriter(sw))
-              println(s"Shutting down due to unexpected error:\n$sw")
-              CoordinatedShutdown.get(system).run(CoordinatedShutdown.ActorSystemTerminateReason)
-            case Success(_) =>
-              CoordinatedShutdown.get(system).run(CoordinatedShutdown.ActorSystemTerminateReason)
-          }
-          Behaviors.same
-        }
-      val system: ActorSystem[Nothing] = ActorSystem[Nothing](guardian, "uexplorer", config)
-      Await.result(system.whenTerminated, Duration.Inf)
-  }
+  override val bootstrap: ZLayer[ZIOAppArgs, Any, Any] =
+    Runtime.setConfigProvider(ExplorerConfig())
 
+  def run =
+    (for {
+      serverFiber   <- Backend.runServer
+      nodePoolFiber <- ZIO.serviceWithZIO[StreamScheduler](_.validateAndSchedule())
+      done = serverFiber.zip(nodePoolFiber)
+    } yield done).provide(
+      ChainIndexerConf.layer,
+      ChainIndexerConf.layer.project(_.mvStore),
+      ChainIndexerConf.layer.project(_.nodePool),
+      MemPool.layer,
+      NodePool.layer,
+      UnderlyingBackend.layer,
+      SttpNodePoolBackend.layer,
+      Backend.layerH2,
+      MetadataHttpClient.layer,
+      BlockHttpClient.layer,
+      PersistentBlockRepo.layer,
+      PersistentBoxRepo.layer,
+      PersistentRepo.layer,
+      StreamScheduler.layer,
+      PluginManager.layer,
+      StreamExecutor.layer,
+      MempoolSyncer.layer,
+      Initializer.layer,
+      BlockReader.layer,
+      BlockWriter.layer,
+      GraphBackend.layer,
+      MvStorage.zlayerWithDefaultDir
```

Comment:
This is really nice to look at :+1: 
Btw you can check out "my" patter for declaring ZLayers https://discord.com/channels/629491597070827530/630498701860929559/1127725301196136528

Reply:
Yup, this is going to be part of the 2. or 3. iteration of migration to Zio :) There is a big list : revisiting logging, layers, etc.

Reply:
Cool. Good luck :crossed_fingers: